### ISSUE TITLE
binary segmentation，acc for background is zero, iou is zero

### ISSUE THREAD

**Issue by weiqingtangx** [open-mmlab/mmsegmentation#1304]
<img width="863" alt="截屏2022-02-19 上午12 33 00" src="https://user-images.githubusercontent.com/5905640/154724022-c8f76b95-84dc-47a0-8cf4-795c80717ff7.png">

Hi, sometime background class IoU and Acc is zero, other time coin class IoU and Acc is zero.

Here is my config file:
`
norm_cfg = dict(type='BN', requires_grad=True)
model = dict(
    type='EncoderDecoder',
    pretrained='open-mmlab://resnet50_v1c',
    backbone=dict(
        type='ResNetV1c',
        depth=50,
        num_stages=4,
        out_indices=(0, 1, 2, 3),
        dilations=(1, 1, 2, 4),
        strides=(1, 2, 1, 1),
        norm_cfg=dict(type='BN', requires_grad=True),
        norm_eval=False,
        style='pytorch',
        contract_dilation=True),
    decode_head=dict(
        type='DepthwiseSeparableASPPHead',
        in_channels=2048,
        in_index=3,
        channels=512,
        dilations=(1, 12, 24, 36),
        c1_in_channels=256,
        c1_channels=48,
        dropout_ratio=0.1,
        num_classes=2,
        norm_cfg=dict(type='BN', requires_grad=True),
        align_corners=False,
        loss_decode=dict(
            type='CrossEntropyLoss', use_sigmoid=False, loss_weight=1.0)),
    auxiliary_head=dict(
        type='FCNHead',
        in_channels=1024,
        in_index=2,
        channels=256,
        num_convs=1,
        concat_input=False,
        dropout_ratio=0.1,
        num_classes=2,
        norm_cfg=dict(type='BN', requires_grad=True),
        align_corners=False,
        loss_decode=dict(
            type='CrossEntropyLoss', use_sigmoid=False, loss_weight=0.4)),
    train_cfg=dict(),
    test_cfg=dict(mode='whole'))
dataset_type = 'PascalVOCDataset'
data_root = 'data/coin'
img_norm_cfg = dict(
    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
crop_size = (512, 512)
train_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='LoadAnnotations'),
    dict(type='Resize', img_scale=(2048, 512), ratio_range=(0.5, 2.0)),
    dict(type='RandomCrop', crop_size=(512, 512), cat_max_ratio=0.75),
    dict(type='RandomFlip', prob=0.5),
    dict(type='PhotoMetricDistortion'),
    dict(
        type='Normalize',
        mean=[123.675, 116.28, 103.53],
        std=[58.395, 57.12, 57.375],
        to_rgb=True),
    dict(type='Pad', size=(512, 512), pad_val=0, seg_pad_val=255),
    dict(type='DefaultFormatBundle'),
    dict(type='Collect', keys=['img', 'gt_semantic_seg'])
]
test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(
        type='MultiScaleFlipAug',
        img_scale=(2048, 512),
        flip=False,
        transforms=[
            dict(type='Resize', keep_ratio=True),
            dict(type='RandomFlip'),
            dict(
                type='Normalize',
                mean=[123.675, 116.28, 103.53],
                std=[58.395, 57.12, 57.375],
                to_rgb=True),
            dict(type='ImageToTensor', keys=['img']),
            dict(type='Collect', keys=['img'])
        ])
]
data = dict(
    samples_per_gpu=4,
    workers_per_gpu=4,
    train=dict(
        type='PascalVOCDataset',
        data_root='data/coin',
        img_dir='JPEGImages',
        ann_dir='SegmentationClass',
        split=['ImageSets/Segmentation/train.txt'],
        pipeline=[
            dict(type='LoadImageFromFile'),
            dict(type='LoadAnnotations'),
            dict(type='Resize', img_scale=(2048, 512), ratio_range=(0.5, 2.0)),
            dict(type='RandomCrop', crop_size=(512, 512), cat_max_ratio=0.75),
            dict(type='RandomFlip', prob=0.5),
            dict(type='PhotoMetricDistortion'),
            dict(
                type='Normalize',
                mean=[123.675, 116.28, 103.53],
                std=[58.395, 57.12, 57.375],
                to_rgb=True),
            dict(type='Pad', size=(512, 512), pad_val=0, seg_pad_val=255),
            dict(type='DefaultFormatBundle'),
            dict(type='Collect', keys=['img', 'gt_semantic_seg'])
        ]),
    val=dict(
        type='PascalVOCDataset',
        data_root='data/coin',
        img_dir='JPEGImages',
        ann_dir='SegmentationClass',
        split='ImageSets/Segmentation/val.txt',
        pipeline=[
            dict(type='LoadImageFromFile'),
            dict(
                type='MultiScaleFlipAug',
                img_scale=(2048, 512),
                flip=False,
                transforms=[
                    dict(type='Resize', keep_ratio=True),
                    dict(type='RandomFlip'),
                    dict(
                        type='Normalize',
                        mean=[123.675, 116.28, 103.53],
                        std=[58.395, 57.12, 57.375],
                        to_rgb=True),
                    dict(type='ImageToTensor', keys=['img']),
                    dict(type='Collect', keys=['img'])
                ])
        ]),
    test=dict(
        type='PascalVOCDataset',
        data_root='data/coin',
        img_dir='JPEGImages',
        ann_dir='SegmentationClass',
        split='ImageSets/Segmentation/val.txt',
        pipeline=[
            dict(type='LoadImageFromFile'),
            dict(
                type='MultiScaleFlipAug',
                img_scale=(2048, 512),
                flip=False,
                transforms=[
                    dict(type='Resize', keep_ratio=True),
                    dict(type='RandomFlip'),
                    dict(
                        type='Normalize',
                        mean=[123.675, 116.28, 103.53],
                        std=[58.395, 57.12, 57.375],
                        to_rgb=True),
                    dict(type='ImageToTensor', keys=['img']),
                    dict(type='Collect', keys=['img'])
                ])
        ]))
log_config = dict(
    interval=50, hooks=[dict(type='TextLoggerHook', by_epoch=False)])
dist_params = dict(backend='nccl')
log_level = 'INFO'
load_from = None
resume_from = None
workflow = [('train', 1)]
cudnn_benchmark = True
optimizer = dict(type='SGD', lr=0.01, momentum=0.9, weight_decay=0.0005)
optimizer_config = dict()
lr_config = dict(policy='poly', power=0.9, min_lr=0.0001, by_epoch=False)
runner = dict(type='IterBasedRunner', max_iters=5)
checkpoint_config = dict(by_epoch=False, interval=1)
evaluation = dict(interval=1, metric='mIoU', pre_eval=True)
work_dir = 'coin'
gpu_ids = [0]
auto_resume = False
`
CLASSES = ('background', 'coin')
PALETTE = [[0, 0, 0], [255, 0, 0]]
Label 
![0a505f87-6b40-f849-9f73-98d8fafa59a6](https://user-images.githubusercontent.com/5905640/154724510-4ad85a63-bfc9-4a9b-a2f2-43fb26b31e9a.png)

I stuck here for one day, could you please give me some help?

**Comment by weiqingtangx** [comment]
> You may need to set the coin's label value as 1. Cause the doc says the label's pixel value range from 0 to N-1. By the way did you resize your label to h*w without channel?

I tried to set label value as 1, however get the same result. Here is my label with value 1.
![0a505f87-6b40-f849-9f73-98d8fafa59a6](https://user-images.githubusercontent.com/5905640/154783890-0ebca967-cb9e-4144-b1c7-b058c3452bd7.png)

**Comment by weiqingtangx** [comment]
> Have you set reduce_zero_label=False in dataset class?

Yes, here is my code. 
    CLASSES = ('background', 'coin')
    PALETTE = [[0, 0, 0], [255, 0, 0]]

    def __init__(self, split, **kwargs):
        super(PascalVOCDataset, self).__init__(
            img_suffix='.jpg', seg_map_suffix='.png', split=split, reduce_zero_label=False, **kwargs)
        assert osp.exists(self.img_dir) and self.split is not None`
`
However it is not working.

**Comment by MengzhangLI** [comment]
Sorry for late reply.

You can modify current config, change dataset to our provided dataset like DRIVE or STARE which also only have one foreground, and train agian. If it works well, if it works well, it is caused by dataset, it not, it may caused by model.

**Comment by weiqingtangx** [comment]
> Sorry for late reply.
> 
> You can modify current config, change dataset to our provided dataset like DRIVE or STARE which also only have one foreground, and train agian. If it works well, if it works well, it is caused by dataset, it not, it may caused by model.

Thank you. I will try it.

**Comment by Senwang98** [comment]
@weiqingtangx 
Have you solve this question?
I meet the same problem, when I use DRIVE dataset to train and using image_demo.py to inference, it can display normally.
When changing to my own dataset, it works bad.
I set n_class=2 and use img[::0] // 128 to make annotation only have 0 or 1.
But I still fail to display.

**Comment by wptoux** [comment]
https://github.com/open-mmlab/mmsegmentation/issues/1335#issuecomment-1060185864

I have met the similar problem, maybe you could set reduce_zero_label to False in LoadAnnotations and try again.

**Comment by Senwang98** [comment]
@wptoux 
reduce_zero_label is default false
